extensions/amp-form/0.1/amp-form.js

```diff
@@ -109,6 +116,39 @@ export class AmpForm {
     }
     this.form_.classList.add(`amp-form-${state}`);
   }
+
+  /**
+   * @param {string} state
+   * @param {!Object} data
+   * @private
+   */
+  renderTemplate_(state, data) {
+    const container = this.form_.querySelector(`[${state}]`);
+    if (container) {
+      // TODO(#3587): Move the cleanup to do during submission.
+      this.cleanupRenderedTemplate_(state);
+      return this.templates_.findAndRenderTemplate(container, data)
+          .then(rendered => {
+            rendered.classList.add(`${state}-message`);
+            container.appendChild(rendered);
+          });
+    }
+  }
+
+  /**
+   * @param {string} state
+   * @private
+   */
+  cleanupRenderedTemplate_(state) {
+    const container = this.form_.querySelector(`[${state}]`);
+    if (!container) {
+      return;
+    }
+    const previousRender = container.querySelector(`.${state}-message`);
```

Comment:
If we do `i-amp-rendered` or such, we can then use a more optimized `dom.childElementByAttr`